### PR TITLE
PROTOCOLS-120 ImapEncoders should no longer be chained

### PR DESCRIPTION
**Pull request by chibenwa**

### CODE REVIEW

File: protocols/imap/src/main/java/org/apache/james/imap/encode/ImapResponseEncoder.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+package org.apache.james.imap.encode;
+
+import java.io.IOException;
+
+import org.apache.james.imap.api.ImapMessage;
+import org.apache.james.imap.api.process.ImapSession;
+
+/**
+ * TODO: correct this API
```
Comment (rouazana) [inline protocols/imap/src/main/java/org/apache/james/imap/encode/ImapResponseEncoder.java]: still TODO?

File: protocols/imap/src/main/java/org/apache/james/imap/encode/ImapResponseEncoder.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+ */
+public interface ImapResponseEncoder<T extends ImapMessage> {
+    Class<T> acceptableMessages();
+
+    /**
+     * Writes response. TODO: pass in writer rather than composer
```
Comment (rouazana) [inline protocols/imap/src/main/java/org/apache/james/imap/encode/ImapResponseEncoder.java]: idem?

File: protocols/imap/src/main/java/org/apache/james/imap/encode/ImapResponseEncoder.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+     * 
+     * @param message
+     *            <code>ImapMessage</code>, not null
+     * @param composer
+     *            <code>ImapResponseComposer</code>, not null
+     * @param session
+     *            TODO
```
Comment (rouazana) [inline protocols/imap/src/main/java/org/apache/james/imap/encode/ImapResponseEncoder.java]: idem?

File: protocols/imap/src/main/java/org/apache/james/imap/encode/main/DefaultImapEncoderFactory.java [hunk 4]
```diff
@@ -48,10 +57,36 @@
...
 import org.apache.james.imap.encode.base.EndImapEncoder;
 
+import com.github.steveash.guavate.Guavate;
+import com.google.common.collect.ImmutableList;
+
 /**
  * TODO: perhaps a POJO would be better
  */
 public class DefaultImapEncoderFactory implements ImapEncoderFactory {
+    private static class DefaultImapEncoder implements ImapEncoder {
```
Comment (rouazana) [inline protocols/imap/src/main/java/org/apache/james/imap/encode/main/DefaultImapEncoderFactory.java]: is the ImapEncoder interface still useful?

File: protocols/imap/src/test/java/org/apache/james/imap/encode/MailboxStatusResponseEncoderTest.java [hunk 5]
```diff
@@ -34,23 +32,17 @@
...
 
     @Before
     public void setUp() throws Exception {
-        mockNextEncoder = mock(ImapEncoder.class);
-        encoder = new MailboxStatusResponseEncoder(mockNextEncoder);
+        encoder = new MailboxStatusResponseEncoder();
     }
 
     @Test
     public void testIsAcceptable() {
```
Comment (rouazana) [inline protocols/imap/src/test/java/org/apache/james/imap/encode/MailboxStatusResponseEncoderTest.java]: you could rename this test

File: protocols/imap/src/test/java/org/apache/james/imap/encode/SearchResponseEncoderTest.java [hunk 6]
```diff
@@ -35,27 +32,19 @@
...
     @Before
     public void setUp() throws Exception {
-        mockNextEncoder = mock(ImapEncoder.class);
         response = new SearchResponse(IDS, null);
-        encoder = new SearchResponseEncoder(mockNextEncoder);
+        encoder = new SearchResponseEncoder();
     }
 
     @Test
     public void testIsAcceptable() {
```
Comment (rouazana) [inline protocols/imap/src/test/java/org/apache/james/imap/encode/SearchResponseEncoderTest.java]: idem

File: protocols/imap/src/main/java/org/apache/james/imap/encode/ImapResponseEncoder.java (hunk 3 above)
Comment (chibenwa) [reply protocols/imap/src/main/java/org/apache/james/imap/encode/ImapResponseEncoder.java]: It's never used, I will remove it in an additional commit...

File: protocols/imap/src/main/java/org/apache/james/imap/encode/main/DefaultImapEncoderFactory.java (hunk 4 above)
Comment (chibenwa) [reply protocols/imap/src/main/java/org/apache/james/imap/encode/main/DefaultImapEncoderFactory.java]: Yes as I need one item to be writing *any* kind of message and not just a single 'sub-class'.

(so do 'unknownResponseEncoder): I'm not able to implement acceptableMessages for them, thus this do not fit the ImapResponseEncoder hence the ImapEncoder separate interface...

**Comment by Arsnael** [comment]
```
[948a192662b75efc8ba6574ff397318af0b35270] [ERROR] /james-project/protocols/imap/src/main/java/org/apache/james/imap/encode/main/DefaultImapEncoderFactory.java:[84,36] method encode in interface org.apache.james.imap.encode.ImapResponseEncoder<T> cannot be applied to given types;
[948a192662b75efc8ba6574ff397318af0b35270]   required: org.apache.james.imap.api.ImapMessage,org.apache.james.imap.encode.ImapResponseComposer
[948a192662b75efc8ba6574ff397318af0b35270]   found: org.apache.james.imap.api.ImapMessage,org.apache.james.imap.encode.ImapResponseComposer,org.apache.james.imap.api.process.ImapSession
[948a192662b75efc8ba6574ff397318af0b35270]   reason: actual and formal argument lists differ in length
```

**Comment by mbaechler** [comment]
Why it's duplicated from https://github.com/linagora/james-project/pull/2961? Did I miss a warning about that?

**Comment by chibenwa** [comment]
Duplicated?

What do you mean?

File: protocols/imap/src/main/java/org/apache/james/imap/encode/ImapResponseEncoder.java [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+
+import java.io.IOException;
+
+import org.apache.james.imap.api.ImapMessage;
+
+public interface ImapResponseEncoder<T extends ImapMessage> {
+    Class<T> acceptableMessages();
+
+    /**
+     * Writes response.
```
Comment (mbaechler) [inline protocols/imap/src/main/java/org/apache/james/imap/encode/ImapResponseEncoder.java]: Why it's named  `encode` then?

File: protocols/imap/src/main/java/org/apache/james/imap/encode/ImapResponseEncoder.java [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+     * 
+     * @param message
```
Comment (mbaechler) [inline protocols/imap/src/main/java/org/apache/james/imap/encode/ImapResponseEncoder.java]: please remove this part of the documentation, information are obvious

**Comment by chibenwa** [comment]
`MemorySpamAssassinContractTest.expungingSpamMessageShouldNotImpactSpamAssassinState » ConditionTimeout`

I'm impatient the fix to be on master!

**Comment by Arsnael** [comment]
test this please

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**